### ISSUE TITLE
Salesforce Hybrid App Deferred Login Error

### ISSUE THREAD

**Issue by dev9163** [forcedotcom/SalesforceMobileSDK-CordovaPlugin#399]
I tried doing deferred login for my connected app. In the bootconfig.json file, I changed the shouldAuthenticate to false.
I then put the following code in the place I want my login to happen:

let oauth = OAuth.createInstance(appId, loginURL, oauthCallbackURL);
    oauth.login().then((oauthData) => {
      DataService.createInstance(oauthData, { proxyURL: "https://dev-cors-proxy.herokuapp.com/" });
    });

All of the params have values. I can run it on the desktop browser with no issues. However, on running it on a device, the login does not happen when required. Instead I get the following error in the console:

`TIMING com.salesforce.oauth:getAuthCredentials: 141.196ms
com.salesforce.util.exec.js:43 TIMING com.salesforce.oauth:getAuthCredentials failed(anonymous function) @ com.salesforce.util.exec.js:43callbackFromNative @ cordova.js:293(anonymous function) @ VM330:1
vendor.js:36880 Never authenticated
vendor.js:54799 EXCEPTION: Uncaught (in promise): Never authenticatedErrorHandler.handleError @ vendor.js:54799IonicErrorHandler.handleError @ vendor.js:135311next @ vendor.js:38484schedulerFn @ vendor.js:39367SafeSubscriber.__tryOrUnsub @ vendor.js:1330SafeSubscriber.next @ vendor.js:1279Subscriber._next @ vendor.js:1232Subscriber.next @ vendor.js:1196Subject.next @ vendor.js:5794EventEmitter.emit @ vendor.js:39353NgZone.triggerError @ vendor.js:27544onHandleError @ vendor.js:27505t.handleError @ polyfills.js:3e.runGuarded @ polyfills.js:3r @ polyfills.js:3i @ polyfills.js:3
vendor.js:54804 ORIGINAL STACKTRACE:ErrorHandler.handleError @ vendor.js:54804IonicErrorHandler.handleError @ vendor.js:135311next @ vendor.js:38484schedulerFn @ vendor.js:39367SafeSubscriber.__tryOrUnsub @ vendor.js:1330SafeSubscriber.next @ vendor.js:1279Subscriber._next @ vendor.js:1232Subscriber.next @ vendor.js:1196Subject.next @ vendor.js:5794EventEmitter.emit @ vendor.js:39353NgZone.triggerError @ vendor.js:27544onHandleError @ vendor.js:27505t.handleError @ polyfills.js:3e.runGuarded @ polyfills.js:3r @ polyfills.js:3i @ polyfills.js:3
vendor.js:54805 Error: Uncaught (in promise): Never authenticated
    at s (polyfills.js:3)
    at s (polyfills.js:3)
    at polyfills.js:3
    at t.invokeTask (polyfills.js:3)
    at Object.onInvokeTask (vendor.js:27475)
    at t.invokeTask (polyfills.js:3)
    at e.runTask (polyfills.js:3)
    at i (polyfills.js:3)ErrorHandler.handleError @ vendor.js:54805IonicErrorHandler.handleError @ vendor.js:135311next @ vendor.js:38484schedulerFn @ vendor.js:39367SafeSubscriber.__tryOrUnsub @ vendor.js:1330SafeSubscriber.next @ vendor.js:1279Subscriber._next @ vendor.js:1232Subscriber.next @ vendor.js:1196Subject.next @ vendor.js:5794EventEmitter.emit @ vendor.js:39353NgZone.triggerError @ vendor.js:27544onHandleError @ vendor.js:27505t.handleError @ polyfills.js:3e.runGuarded @ polyfills.js:3r @ polyfills.js:3i @ polyfills.js:3`

Can someone please help me with this. It's crashing my brains out since a long time now :(

**Comment by bhariharan** [comment]
The error looks like it's here (outside the `OAuth` plugin):
```
EXCEPTION: Uncaught (in promise): Never authenticatedErrorHandler.handleError @ vendor.js:54799IonicErrorHandler.handleError @ vendor.js:135311next @ vendor.js:38484schedulerFn @ vendor.js:39367SafeSubscriber.__tryOrUnsub @ vendor.js:1330SafeSubscriber.next
```
Have you tried our generated app with deferred authentication out of the box? This forum is reserved for known issues with our code. Please post other issues/queries in our community [here](https://plus.google.com/u/1/communities/114225252149514546445?cfem=1).

**Comment by dev9163** [comment]
Have you tried our generated app with deferred authentication out of the box?

Can you please help me with the link for the same?
@bhariharan

**Comment by bhariharan** [comment]
Generate an app using our CLI [forcehybrid](https://www.npmjs.com/package/forcehybrid). The CLI has easy-to-use step-by-step instructions.

**Comment by dev9163** [comment]
This app does not have deferred login. I tried making it deferred, but I was not successful, I am using ionic 2 and angular 2. The deferred login I see is for angular 1. Can you help me with the snippet for deferred login in angular 2? The code I have done throws never authenticated error by cordova.

**Comment by jaygauten** [comment]
hi @bhariharan, can you guide us how to create a sample deferred login? also it seems the guide here: https://developer.salesforce.com/docs/atlas.en-us.noversion.mobile_sdk.meta/mobile_sdk/hybrid_deferred_login.htm is kinda confusing

**Comment by bhariharan** [comment]
@jaygauten The dev guide explains it pretty well. Here are the steps, in sequence:
- Change the value of `shouldAuthenticate` in your app's `bootconfig.json` to `false`.
- This will load the start page directly.
- Assuming you have a login button, clicking this will trigger the login flow. The click handler for this button must call our `OAuth` plugin's `authenticate` method. If it simply redirects to the login page, this will not work in the container's context.

**Comment by jaygauten** [comment]
I did create a fresh app generated using forcehybrid then followed the steps, what happen though instead of showing the startpage it still shows the login page with the back button enabled.

Tested using iOS

**Comment by wmathurin** [comment]
The bug is fixed for the upcoming release (7.1). Here is a workaround for it: https://github.com/forcedotcom/SalesforceMobileSDK-Templates/issues/179

**Comment by jaygauten** [comment]
thanks @wmathurin @bhariharan
